src/Symfony/Component/HttpKernel/ControllerMetadata/ControllerMetadataUtil.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpKernel\ControllerMetadata;
+
+use Doctrine\Common\Util\ClassUtils;
+
+abstract class ControllerMetadataUtil
+{
+    /**
+     * Gets a class name (if available) and method/function name from a callable.
+     *
+     * @param callable $controller
+     *
+     * @return string[]|null the class name and method name (where applicable)
+     */
+    public static function getControllerLogicalName(callable $controller)
+    {
+        if ($controller instanceof \Closure) {
+            // cannot store any metadata of anonymous functions
+            return null;
+        }
+
+        if (is_object($controller)) {
+            // callable class
+            $controller = array($controller, '__invoke');
+        } elseif (is_string($controller)) {
+            // normal function
+            $controller = array(null, $controller);
+        }
+
+        $className = null;
+        $method = $controller[1];
+
+        if (null !== $controller[0]) {
+            if (!is_string($controller[0])) {
+                $className = class_exists('Doctrine\Common\Util\ClassUtils') ? ClassUtils::getClass($controller[0]) : get_class($controller[0]);
```

Comment:
As long as it's okay with the codestyle, I've (shamelessly) copied it from somewhere else. Main reason to not write the constant notation was that I wasn't sure if it would trigger an error. I forgot that class_exists suppresses this and simply works: https://3v4l.org/0BVVI